numpyro/infer/svi.py

```diff
@@ -137,6 +144,46 @@ def update(self, svi_state, *args, **kwargs):
         loss_val, optim_state = self.optim.eval_and_update(loss_fn, svi_state.optim_state)
         return SVIState(optim_state, rng_key), loss_val
 
+    def run(self, rng_key, num_steps, *args, progress_bar=True, **kwargs):
+        """
+        (EXPERIMENTAL INTERFACE) Run SVI with `num_steps` iterations, then return
+        the optimized parameters and the stacked losses at every step. If `num_steps`
+        is large, setting `progress_bar=False` can make the run faster.
+
+        .. note:: For a complex training process (e.g. the one requires early stopping,
+            epoch training, varying args/kwargs,...), we recommend to use the more
+            flexible methods :meth:`init`, :meth:`update`, :meth:`evaluate` to
+            customize your training procedure.
+
+        :param jax.random.PRNGKey rng_key: random number generator seed.
+        :param int num_steps: the number of optimization steps.
+        :param args: arguments to the model / guide
+        :param bool progress_bar: Whether to enable progress bar updates. Defaults to
+            ``True``.
+        :param kwargs: keyword arguments to the model / guide
+        :return: a namedtuple with fields `params` and `losses` where `params`
+            holds the optimized values at :class:`numpyro.param` sites,
+            and `losses` is the collected loss during the process.
+        :rtype: SVIRunResult
+        """
+        def body_fn(svi_state, carry):
+            svi_state, loss = self.update(svi_state, *args, **kwargs)
+            return svi_state, loss
+
+        svi_state = self.init(rng_key, *args, **kwargs)
+        if progress_bar:
```

Comment:
I think that for SVI printing the loss at some regular frequency is quite useful to see the trend - is the loss going down, how fast is it going down and whether it has converged. Using a progress bar might not be as useful in that regard since the trend isn't quite as clear. If we are to use a progress bar, how about adding % decrease from initial loss value ("x% decrease") along with the loss to help visualize the trend.

Reply:
I agree that more diagnostic infos would be pretty helpful here.

This requires more control values for the report (or we can hard-code them to some reasonable values). Currently, I think that we can return the average loss (over the last ... steps) to account for the fluctuations. About `% decrease`, should we record it with respect to some checkpoint `iter` (initial losses are usually pretty high)?

Reply:
> This requires more control values for the report (or we can hard-code them to some reasonable values). Currently, I think that we can return the average loss (over the last ... steps) to account for the fluctuations.

I think we can just refresh the bar less frequently or do some exponential smoothing if the fluctuations are too high. 

> About % decrease, should we record it with respect to some checkpoint iter (initial losses are usually pretty high)?

It should be in the log scale so can't be too bad? I have also seen problems, specially with bad initialization, where it doesn't go down much at all.

Reply:
> we can just refresh the bar less frequently

I think we should do this. Currently, I see the losses change so fast that I can't keep up with them.

> specially with bad initialization

Haha, I have seen the reverse more: bad initialization will have a very high loss, then it dropped down like 1e6 to 100... XD

Just to confirm your recommendation, "x% decrease" is computed as: (losses[-1] / losses[0]) * 100? What happens when `loss` is negative?

Reply:
> "x% decrease" is computed as: (losses[-1] / losses[0]) * 100? What happens when loss is negative?

I meant ((losses[0] - losses[-1])/losses[0]) * 100. This should be a positive quantity if the optimization is working as expected (it can on occasion go negative I think but should be rare; in any case, we shouldn't change what we display).

> bad initialization will have a very high loss, then it dropped down like 1e6 to 100.

In cases like this, it will start off with 0% decrease and rapidly go to 99% decrease. That might not be super useful but using some other loss value as baseline will be harder to explain. Another thing we could do is to just mention the initial loss in the progress bar instead of reporting a percent decrease if you think that will be more generally useful, or just not have a progress bar at all and report loss after every 5% progress.
```
[5%  -- Iter: 200] Loss: 100
[10% -- Iter: 400] Loss: ...
```

Reply:
> Another thing we could do is to just mention the initial loss

This looks more readable to me. And I think it is good idea to report loss every 5% progress. I also prefer reporting average loss of the last 5% progress. Let me see if I can integrate all those information in progbar.

> just not have a progress bar

I think looking at progbar slowly runs is more fun than waiting for the next print. :D

Reply:
@neerajprad I just updated the progbar. Currently, it is like this
```
100%|██████████████████████████████████████████████| 10000/10000 [00:11<00:00, 890.02it/s, init loss: 8.6413, avg. loss [9501-10000]: -1.7689]
```